libevmasm/SemanticInformation.cpp

```diff
@@ -278,12 +291,63 @@ bool SemanticInformation::invalidatesStorage(Instruction _instruction)
 	case Instruction::CREATE:
 	case Instruction::CREATE2:
 	case Instruction::SSTORE:
-		return true;
+		return yul::SideEffects::Write;
+
+	case Instruction::SLOAD:
+	case Instruction::STATICCALL:
+		return yul::SideEffects::Read;
+
 	default:
-		return false;
+		return yul::SideEffects::None;
+	}
+}
+
+yul::SideEffects::Effect SemanticInformation::otherState(Instruction _instruction)
+{
+	switch (_instruction)
+	{
+	case Instruction::CALL:
+	case Instruction::CALLCODE:
+	case Instruction::DELEGATECALL:
+	case Instruction::CREATE:
+	case Instruction::CREATE2:
+	case Instruction::SELFDESTRUCT:
+		// Strictly speaking, log0, .., log4 writes to the state, but the EVM cannot read it, so they
+		// are just marked as having 'other side effects.'
+		return yul::SideEffects::Write;
+
+	case Instruction::EXTCODESIZE:
+	case Instruction::EXTCODEHASH:
+	case Instruction::RETURNDATASIZE:
+	case Instruction::BALANCE:
+	case Instruction::SELFBALANCE:
+	case Instruction::RETURNDATACOPY:
+	case Instruction::EXTCODECOPY:
+	case Instruction::STATICCALL:
+		// PC and GAS are specifically excluded here. Instructions such as CALLER, CALLVALUE,
+		// ADDRESS are excluded because they cannot change during execution.
+		return yul::SideEffects::Read;
+
+	default:
+		return yul::SideEffects::None;
 	}
 }
 
+yul::SideEffects::Effect SemanticInformation::otherImmovables(Instruction _instruction)
```

Comment:
I don't think we should have this as a none/read/write setting.

The for loop optimizer used to use `movable` as the only criterion. Now we are allowing some opcodes that used to be non-movable. Maybe we should introduce `movableApartFromEffects`?

Reply:
And then the criterion for the for loop is `movableApartFromEffects && effects-do-not-conflict-with-loop-body`

Reply:
Is this how you want it to look like:

```cpp
bool SemanticInformation::movableApartFromEffects(Instruction _instruction)
{
	switch (_instruction)
	{
	case Instruction::EXTCODEHASH:
	case Instruction::EXTCODESIZE:
	case Instruction::RETURNDATASIZE:
	case Instruction::BALANCE:
	case Instruction::SELFBALANCE:
	case Instruction::SLOAD:
		return true;

	default:
		return movable(_instruction);
	}
}

```

I like it because it makes it explicit what additional instructions qualify in the new optimization.

Reply:
I'll have to check the exact list, but yes!